### DISCUSSION TITLE
Flac/Mp3/Amazon/SiriusXM/Home Assistant

### DISCUSSION THREAD

**Discussion by bs42** [micro-nova/AmpliPi#65] [Ideas]
I would love to see the ability to plug in USB Storage and play Flac/Mp3 files from that source.  Perhaps something like Mopidy.  The ability to control that via the API would be great.

Amazon Music HD and SiriusXM would be great options too, however I understand that's less likely due to those companies not being very open with their access.  (A connected Sonos Port will work for now)

As soon as I can get my hands on the device I plan to work on a module for [Home Assistant](https://home-assistant.io) to control this thing, unless it's done before then.

**Comment by linknum23** [comment]
Mopidy looks like a great solution for playing local files from usb attached storage, the best part is there are existing apps that can be used to select and play songs.  Feel free to add an issue for Mopidy integration. 

Based on my initial research sirius xm and Amazon music are not currently supported on raspbian. These are definitely something to look into for the future, though I would not expect direct support for them in the near term. For now these would need to be supported with an external device like the sonos mentioned.

Also we would love your help creating a plugin for home assistant! I hope our api can make development straight forward. What are you thinking of automating initially? I've got some ideas of how I would use this, but would love to hear what your plans are, so we can keep them in mind for future development.

**Reply by bs42** [reply]
I'll put in the Mopidy request shortly.

As for Home Assistant, I'd like to ensure it can do everything the existing Sonos plugin can do.  Control/assign input/output groups, mute/unmute/volume making it a "media_player" in Home Assistant .  Your REST API looks easy enough to work with.

I would like to suggest something though, and I understand it's potentially a larger undertaking:  websocket support.

The idea is that Home Assistant (or another home automation system) could maintain a constant connection to the AmpliPi in order to receive real time updates of play/pause/volume/group/source/etc information in order for the system to be always update without any latency from the delay in waiting for whatever interval the REST api was set to poll at.  Potentially reducing CPU cycles for the AmpliPi as well since it's not constantly processing API requests for information it could have just pushed.

**Reply by kaikreuzer** [reply]
> could maintain a constant connection to the AmpliPi in order to receive real time updates

Working on openHAB, I can only support this request - for HA systems it is very desirable to receive changes automatically instead of having to poll them. Besides websockets, Server Sent Events (SSE) can also be a suitable technology, which can nicely complement a REST API.

**Reply by linknum23** [reply]
It looks like SSE might make more sense initially to add since it appears complementary to the rest api. Having this would fix some outstanding issues in the webapp. Are there any downsides to using SSE instead of websockets?

**Reply by bs42** [reply]
Either method would accomplish the same end result in my opinion, so whatever works best with the existing code base I think makes the most sense.

**Reply by kaikreuzer** [reply]
FWIW, we use SSE in openHAB and it works very stable and reliably. Websockets additionally bring bi-directional communication, but that's a feature you usually do not want/need as this is already covered by the REST endpoints.

**Reply by linknum23** [reply]
Just added #73, thanks!

**Reply by rickballard** [reply]
A USB-adaptor can be used to create a mirco-SD memory card slot, inside the port.  

So, for example, you could store event-triggered MP3's in there.  If your motion sensor detects the raising of a dog's hind leg, it could trigger a file that broadcasts: "Get off my bloody lawn!" Meanwhile, if your wireless camera sees someone it doesn't recognize, it could trigger a machine gun sound, which can be quite handy in some parts of Los Angeles.  

However, you could also ask Google Assistant, Alexa, etc, to broadcast a customized message to each kind of intruder, right...?

**Reply by linknum23** [reply]
Any sort of it this that style integration would probably make sense to be part of an existing home automation framework such as OpenHAB, HomeAssistant. We have plans to integrate with both of those frameworks.

**Reply by rickballard** [reply]
So, you we can natively HOST the Home Assistant server in the AmpliPi.

After all, the AmpliPi is basically a Rasberry Pi with a Bunch of Amps (BoA), and the Pi is the preferred a platform for smart home frameworks.  AmpliPi BoA-functionality addresses a deeply-niched -- and fading -- audiophile market, while AmpliPi Server functionality addresses a market that is broad-based and exploding.  It would thus be naïve to launch the AmpliPi without a dedicated container for the Home Assistant server.  

If you fail to set the direction for your future developer community, (who will mostly be smart home integrators), you will find customer support becomes increasingly problematic.  This future community polarize, killer-endorsers or killer-critics; they either support your market -- and your marketing efforts -- or they break it. 

Nobody will want the AmpliPi when the AmpliPi+ is available, (the version with the Home Assistant container), especially since it will be the same price.  So, pretending that the AmpliPi can stand alone with only its BoA-functionality, would be a disasterous marketing strategy.  The first "smarts" developer to upgrade it would deprecate your launch market and the AmpliPi would be OWL -- Obsolete When Launched.  

If it is already too late to save the reputation of your exalted marketing strategists....  Perhaps you should post a style guide for the Home Assistant hooks in your mobile app....  cue the empathetic groan....  :(

**Reply by mabowers** [reply]
Hi @rickballard, thanks for your feedback. In response to your points:

> So, you we can natively HOST the Home Assistant server in the AmpliPi.

Yes, if a user wants to host Home Assistant (or OpenHAB, or any other home automation software) directly on their AmpliPi, they are more than welcome to do so. The AmpliPi box contains a standard Raspberry Pi CM3+ that users have full control over. Even with the AmpliPi software and multiple streams running, there is plenty of extra compute power to handle other applications (like home automation software).

While some users will do that, we think many will continue to run their home automation setup on an existing server or appliance (e.g. apps server, NAS, dedicated home automation Pi, etc.) and interact with AmpliPi API over the network.

> After all, the AmpliPi is basically a Rasberry Pi with a Bunch of Amps (BoA), and the Pi is the preferred a platform for smart home frameworks. AmpliPi BoA-functionality addresses a deeply-niched -- and fading -- audiophile market, while AmpliPi Server functionality addresses a market that is broad-based and exploding. It would thus be naïve to launch the AmpliPi without a dedicated container for the Home Assistant server.
>
> If you fail to set the direction for your future developer community, (who will mostly be smart home integrators), you will find customer support becomes increasingly problematic. This future community polarize, killer-endorsers or killer-critics; they either support your market -- and your marketing efforts -- or they break it.
> 

The core mission of AmpliPi is to be a great multi-zone audio controller, and that is where we want to keep our focus.

Our plan with AmpliPi is to be able to integrate with a wide variety of streaming and control sources, and to provide a flexible, well-documented API for users to easily integrate AmpliPi with other software and systems. We do **not** plan to marry AmpliPi to any particular home automation software, nor to position AmpliPi as an all-in-one home automation box.

> Nobody will want the AmpliPi when the AmpliPi+ is available, (the version with the Home Assistant container), especially since it will be the same price. So, pretending that the AmpliPi can stand alone with only its BoA-functionality, would be a disasterous marketing strategy. The first "smarts" developer to upgrade it would deprecate your launch market and the AmpliPi would be OWL -- Obsolete When Launched.
> 

The AmpliPi system is open, extendable, and customizable. Users can write their own integrations and plugins, or make use of those developed by the community. As the core AmpliPi software is updated, existing users can take advantage of those updates (and unlike some [other home audio vendors](https://www.engadget.com/2019-12-31-sonos-recycle-mode-explanation-falls-flat.html), we never plan on killing off perfectly working units in the field.) As the AmpliPi hardware evolves (e.g. move to CM4 in the future), we even plan on offering DIY upgrade kits for existing AmpliPi customers. AmpliPi is anything but "obsolete when launched."

(Just to be clear: we don't offer a product called "AmpliPi+" and there are no current plans to do so.)

> If it is already too late to save the reputation of your exalted marketing strategists.... Perhaps you should post a style guide for the Home Assistant hooks in your mobile app.... cue the empathetic groan.... :(

We always welcome feedback – but this kind of sarcastic tone in your last few sentences is needlessly negative, and doesn't add anything to the conversation. We would please like to keep this a [respectful, constructive environment](https://docs.github.com/en/github/site-policy/github-community-guidelines#building-a-strong-community) :)

**Reply by bs42** [reply]
@mabowers there are those of us who've backed you and are grateful for what the AmpliPi is positioned to be.  Please don't let weird negative comments such as this get you down.  The rest of us are very excited to get this product in our hands (and in our equipment closets).

I, for one, do not wish to have my home automation software running on my audio equipment, I have a NUC for that purpose.

Your designed extensibility and open software/hardware is exactly why I backed the project.  Some people may enjoy giving Sonos stacks of cash for an inferior setup but I'm looking forward to putting some speakers in my ceiling and attaching the AmpliPi to it.  June (or whenever done) can't come soon enough.
